http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BackgroundProcess.scala

```diff
@@ -0,0 +1,92 @@
+package org.broadinstitute.dsde.workbench.leonardo.monitor
+
+import cats.Show
+import cats.effect.Async
+import cats.syntax.all._
+import fs2.Stream
+import org.broadinstitute.dsde.workbench.model.TraceId
+import org.broadinstitute.dsde.workbench.openTelemetry.OpenTelemetryMetrics
+import org.typelevel.log4cats.StructuredLogger
+
+import java.time.Instant
+import java.util.UUID
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * Defines a common interface that certain background processes should implement.
+ * This type of background processes do the following periodically:
+ * 1. Query the database for a list of candidates
+ * 2. Filter the candidates based on some criteria
+ * 3. Perform some action on the filtered candidates
+ */
+trait BackgroundProcess[F[_], A] {
+  def monitorType: String
+  def interval: FiniteDuration
+
+  /**
+   * Given a candidate, determine whether it should be filtered out.
+   * true: keep the candidate
+   * false: filter out the candidate
+   */
+  def filterCriteria(a: A, now: Instant)(implicit
+    F: Async[F],
+    metrics: OpenTelemetryMetrics[F],
+    logger: StructuredLogger[F]
+  ): F[Boolean]
+
+  /**
+   * Get a collection of candidates from the database
+   */
+  def dbSource(): F[Seq[A]]
+
+  /**
+   * Perform some action on the candidate
+   */
+  def action(a: A, traceId: TraceId, now: Instant)(implicit F: Async[F]): F[Unit]
+
+  /**
+   * The main logic of the background process at each interval
+   */
+  private def check(implicit
+    showA: Show[A],
+    F: Async[F],
+    metrics: OpenTelemetryMetrics[F],
+    logger: StructuredLogger[F]
+  ): F[Unit] = for {
+    now <- F.realTimeInstant
+    _ <- logger.info(
+      s"[logAlert/${monitorType}Heartbeat] Executing ${monitorType} check at interval of ${interval}"
+    )
+    candidates <- dbSource()
+    filteredA <- candidates.toList.filterA { a =>
+      filterCriteria(a, now)
+    }
+    _ <- metrics.gauge(s"${monitorType}/numOfRuntimes",
```

Comment:
Is a gauge the right metric to collect here? Would a direct count be easier? I know the current gauge metrics on autopause runtimes on Grafana is not easy to interpret because of that

Reply:
I just left everything what it was...didn't want to mess with any of these since I know ppl are actively looking at things still :)

Reply:
Why not abstract the metrics and logging side effects to their own method so that `check()` is generic?

Reply:
I want to make sure metrics are logged at one place and in a consistent manner so that individual "monitors" don't have to worry about this...